runtime/calamari/tests/xcm_tests.rs

```diff
@@ -0,0 +1,1791 @@
+		is_frozen: false,
+		is_sufficient: true,
+	};
+
+	// Register ParaA native asset in ParaA
+	ParaA::execute_with(|| {
+		assert_ok!(AssetManager::register_asset(
+			parachain::Origin::root(),
+			// This need to be changed starting from v0.9.16
+			// need to use something like MultiLocation { parents: 0, interior: here} instead
+			source_location.clone(),
+			asset_metadata.clone()
+		));
+		assert_ok!(AssetManager::set_units_per_second(
+			parachain::Origin::root(), 
+			a_currency_id, 
+			0u128));
+		assert_eq!(
+			Some(a_currency_id),
+			AssetManager::location_asset_id(source_location.clone())
+		);
+	});
+
+	ParaB::execute_with(|| {
+		assert_ok!(AssetManager::register_asset(
+			parachain::Origin::root(),
+			source_location.clone(),
+			asset_metadata.clone()
+		));
+		assert_ok!(AssetManager::set_units_per_second(
+			parachain::Origin::root(), 
+			a_currency_id, 
+			units_per_second));
+		assert_eq!(
+			Some(a_currency_id),
+			AssetManager::location_asset_id(source_location.clone())
+		);
+	});
+
+	let dest = MultiLocation {
+		parents: 1,
+		interior: X2(
+			Parachain(2),
+			AccountId32 {
+				network: NetworkId::Any,
+				id: ALICE.into(),
+			},
+		),
+	};
+
+	// Transfer ParaA balance to B
+	ParaA::execute_with(|| {
+		assert_ok!(parachain::XTokens::transfer_with_fee(
+			parachain::Origin::signed(ALICE.into()),
+			parachain::CurrencyId::MantaCurrency(a_currency_id),
+			amount,
+			1,
+			Box::new(VersionedMultiLocation::V1(dest)),
+			dest_weight,
+		));
+		assert_eq!(
+			parachain::Balances::free_balance(&ALICE.into()),
+			INITIAL_BALANCE - amount - fee
+		)
+	});
+
+	ParaB::execute_with(|| {
+		assert_eq!(parachain::Assets::balance(a_currency_id, &ALICE.into()), amount);
+	});
+}
+
+#[test]
+fn send_para_a_asset_from_para_b_to_para_c_with_trader() {
+	MockNet::reset();
+
+	// source location of para a asset
+	let source_location = AssetLocation(VersionedMultiLocation::V1(MultiLocation::new(
+		1,
+		X2(Parachain(1), PalletInstance(PALLET_BALANCES_INDEX)),
+	)));
+	let a_currency_id = 0u32;
+	let mut amount = 888u128;
+	let units_per_second = 1_250_000u128;
+	let dest_weight = 800_000u64;
+	let fee = calculate_fee(units_per_second, dest_weight);
+
+	let asset_metadata = parachain::AssetRegistarMetadata {
+		name: b"ParaAToken".to_vec(),
+		symbol: b"ParaA".to_vec(),
+		decimals: 18,
+		evm_address: None,
+		min_balance: 1,
+		is_frozen: false,
+		is_sufficient: true,
+	};
+
+	// register a_currency in ParaA, ParaB and ParaC
+
+	// we don't charge any fee in A
+	ParaA::execute_with(|| {
+		assert_ok!(parachain::AssetManager::register_asset(
+			parachain::Origin::root(),
+			// we need to change this on/after v0.9.16
+			source_location.clone(),
+			asset_metadata.clone()
+		));
+		assert_ok!(AssetManager::set_units_per_second(
+			parachain::Origin::root(), 
+			a_currency_id, 
+			0u128));
+		assert_eq!(
+			Some(a_currency_id),
+			parachain::AssetManager::location_asset_id(source_location.clone())
+		);
+	});
```

Comment:
There are many similar blocks like this, can we refactor this to a function to make the tests shorter to read?

Reply:
Not sure how to do that honestly. It need to be a closure to capture the context.

Reply:
This can be a good first issue for someone new actually. To dedupe and comment this code more.

Reply:
I mean the blocks that do almost exactly this code,

```rust
fn assert_register<P>(source_location: ?, asset_metadata: ?, currency_id: ?) 
where
    P: ?
{
     P::execute_with(|| {
        assert_ok!(parachain::AssetManager::register_asset(
            parachain::Origin::root(),
            source_location.clone(),
            asset_metadata.clone(),
        ));
        assert_ok!(AssetManager::set_units_per_second(
            parachain::Origin::root(),
            currency_id,
            0
        ));
        assert_eq!(
            Some(currency_id),
            parachain::AssetManager::location_asset_id(source_location.clone())
        );
    })
}
```

Reply:
I think it worth-while to make the tests more abstract. But let's leave that to the next PR.